liblangutil/SourceLocation.h

```diff
@@ -69,35 +71,41 @@ struct SourceLocation
 
 	int start;
 	int end;
-	std::shared_ptr<std::string const> sourceName;
+	std::shared_ptr<CharStream> source;
 };
 
 /// Stream output for Location (used e.g. in boost exceptions).
 inline std::ostream& operator<<(std::ostream& _out, SourceLocation const& _location)
 {
 	if (_location.isEmpty())
 		return _out << "NO_LOCATION_SPECIFIED";
-	return _out << *_location.sourceName << "[" << _location.start << "," << _location.end << ")";
+
+	if (_location.source.get())
+		_out << _location.source->name();
+
+	_out << "[" << _location.start << "," << _location.end << ")";
+
+	return _out;
 }
 
 bool SourceLocation::operator<(SourceLocation const& _other) const
 {
-	if (!sourceName || !_other.sourceName)
-		return std::make_tuple(int(!!sourceName), start, end) < std::make_tuple(int(!!_other.sourceName), _other.start, _other.end);
+	if (!source|| !_other.source)
+		return std::make_tuple(int(!!source), start, end) < std::make_tuple(int(!!_other.source), _other.start, _other.end);
 	else
-		return std::make_tuple(*sourceName, start, end) < std::make_tuple(*_other.sourceName, _other.start, _other.end);
+		return std::make_tuple(source->name(), start, end) < std::make_tuple(_other.source->name(), _other.start, _other.end);
 }
 
 bool SourceLocation::contains(SourceLocation const& _other) const
 {
-	if (isEmpty() || _other.isEmpty() || ((!sourceName || !_other.sourceName || *sourceName != *_other.sourceName) && (sourceName || _other.sourceName)))
+	if (isEmpty() || _other.isEmpty() || ((!source || !_other.source || source->name() != _other.source->name()) && (source || _other.source)))
```

Comment:
Again, `source.get() != _outher.source.get()` instead of comparing source names.